crates/ra_hir/src/ty/tests.rs

```diff
@@ -2319,3 +2319,27 @@ fn typing_whitespace_inside_a_function_should_not_invalidate_types() {
         assert!(!format!("{:?}", events).contains("infer"), "{:#?}", events)
     }
 }
+
+#[test]
+fn no_such_field_diagnostics() {
+    let diagnostics = MockDatabase::with_files(
+        r"
+        //- /lib.rs
+        struct S { foo: i32, bar: () }
+        impl S {
+            fn new() -> S {
+                S {
+                    foo: 92,
+                    baz: 62,
+                }
+            }
+        }
+        ",
+    )
+    .diagnostics();
+
+    assert_snapshot_matches!(diagnostics, @r###"
+"baz: 62": no such field
```

Comment:
is this the right range for this error? wouldn't it be better to have it just on the `baz`?

Reply:
This is the right `SourceNode` for error. The highlighting range(es) should ideally be specified separately. That is, we'll extend the diagnostics trait, but I want to start with a minimal API first. FWIW, the API is very much inspired by https://github.com/JetBrains/kotlin/blob/master/compiler/frontend/src/org/jetbrains/kotlin/diagnostics/Diagnostic.java :)